modules/transactions/src/main/java/org/apache/ignite/internal/tx/impl/TxManagerImpl.java

```diff
@@ -530,6 +558,41 @@ private TxStateMeta coordinatorFinalTxStateMeta(boolean commit, HybridTimestamp
         return new TxStateMeta(commit ? COMMITED : ABORTED, localNodeId.get(), commitTimestamp);
     }
 
+    /**
+     * Check whether previously enlisted primary replicas aren't expired and that commit timestamp is less or equal than primary replicas
+     * expiration timestamp. Given method will either complete result future with void or {@link PrimaryReplicaExpiredException}
+     *
+     * @param enlistedGroups enlisted primary replicas map from groupId to enlistment consistency token.
+     * @param commitTimestamp Commit timestamp.
+     * @return Verification future.
+     */
+    private CompletableFuture<Void> verifyCommitTimestamp(Map<TablePartitionId, Long> enlistedGroups, HybridTimestamp commitTimestamp) {
+        var verificationFutures = new CompletableFuture[enlistedGroups.size()];
+        int cnt = -1;
+
+        for (Map.Entry<TablePartitionId, Long> enlistedGroup : enlistedGroups.entrySet()) {
+            TablePartitionId groupId = enlistedGroup.getKey();
+            Long expectedEnlistmentConsistencyToken = enlistedGroup.getValue();
+
+            verificationFutures[++cnt] = placementDriver.getPrimaryReplica(groupId, commitTimestamp)
+                    .thenAccept(currentPrimaryReplica -> {
+                        if (currentPrimaryReplica == null
+                                || !expectedEnlistmentConsistencyToken.equals(currentPrimaryReplica.getStartTime().longValue())
+                                || commitTimestamp.after(currentPrimaryReplica.getExpirationTime())
```

Comment:
Is it really possible when you pass a commit timstamp in getPrimaryReplica method?

Reply:
According to the current implementation - no. However, it seems more reliable and clear to explicitly check it. So, I'd rather left this as is.

Reply:
Why do you think it is not possible? This is a crucial part: can I trust that any (non-local) replica is primary or not in a particular moment just by using the placement driver API.

Reply:
Because getPrimaryReplica(..., commitTimestamp) won't return expired lease
`if (lease.isAccepted() && lease.getExpirationTime().after(timestampWithClockSkew)) {`
thus it's not possible. So yep, it's possible to change given commitTimestamp check with an assert. However I believe that current approach is better cleaner.

Reply:
Why does we consider a clock skew in placement driver API (and when I looked in the implementation we doing it twice)? I think we should describe this case in detail.
I always think, that the lease expiration date grater than lease start date, it is a contract, but here we are checking it....

Reply:
I've removed the "twice" CLOCK_SKEW addition. And we use CLOCK_SKEW related logic because it's a contract of how await and get works.